alica_engine/include/engine/blackboard/Blackboard.h

```diff
@@ -1,47 +1,325 @@
+    {
+        const char* type = getTypeName<T>();
+        const auto yamlTypeIt = _yamlType.find(key);
+        if (yamlTypeIt == _yamlType.end()) {
+            // key is not found in yaml file
+            if (!type) {
+                // T is an unknown type, use std::any
+                _vals[key] = std::any{std::forward<T>(value)};
+            } else {
+                // T is a known type, just assign (type conversions are taken into account)
+                _vals[key] = std::forward<T>(value);
+            }
+        } else {
+            // key is found in yaml file
+            const auto& yamlType = yamlTypeIt->second;
+            if (yamlType == "std::any") {
+                // yaml type is std::any, so can store values of any type, therefore use std::any for variant type
+                _vals[key] = std::any{std::forward<T>(value)};
+            } else {
+                if (!type) {
+                    // T is an unknown type, but yaml type is always known
+                    throw BlackboardException(stringify("set() failure, unknown type used to set key: ", key, ", expected type (from yaml file): ", yamlType));
+                } else if (yamlType != type) {
+                    // T is a known type, but type mismatch
+                    throw BlackboardException(
+                            stringify("set() type mismatch, key: ", key, ", set type: ", type, ", expected type (from yaml file): ", yamlType));
+                } else {
+                    // T is a known type & types match, assign
+                    _vals[key] = std::forward<T>(value);
+                }
+            }
+        }
+    }
+
+    void map(const std::string& srcKey, const std::string& targetKey, const BlackboardImpl& srcBb)
+    {
+        // Note: srcKey & targetKey has to be set & both have to be found in the yaml (pml or beh) file
+        // mapping succeeds as long as targetType is constructible from srcType (so conversions are supported)
+        std::visit(
+                [srcKey, targetKey, this](auto&& srcValue) {
+                    // set the target as a side effect, throws if targetType is not constructible from srcType
+                    _vals.at(targetKey) = makeBBValueForIndex<false>::make(_vals.at(targetKey).index(), std::forward<decltype(srcValue)>(srcValue));
+                },
+                srcBb._vals.at(srcKey));
+    }
+
+    bool hasValue(const std::string& key) const { return _vals.count(key); }
+    void removeValue(const std::string& key) { _vals.erase(key); }
+
+    void clear() { _vals.clear(); }
+    bool empty() const { return _vals.empty(); }
+    size_t size() const { return _vals.size(); }
+
+private:
+    std::unordered_map<std::string, BBValueType> _vals;
+    std::unordered_map<std::string, std::string> _yamlType;
+
+    friend BlackboardUtil;
+
+    template <class T>
+    static const char* getTypeName()
+    {
+        auto typeIndex = BBValueType{T{}}.index() - 1;
```

Comment:
we can call `getTypeName(typeIndex)` here to avoid code duplication. Because getTypeName(std::size_t index) already accounts for monostate, you dont have to substract 1 from typeIndex then.